### ISSUE TITLE
Multi Step LSTM using PPO

### ISSUE THREAD

**Issue by AntreasAntoniou** [tensorforce/tensorforce#304]
Hello guys,

I am trying to implement the paper "Neural Optimizer Search via RL". I have implemented the majority of the network, leaving the RL agent which I have little experience of building for last. Your framework seems to have the required technique they used in the paper. The RL algorithm is the PPO, but they use an RNN to predict items to combine to generate an optimizer for a lower level task. But, in essence, the high level task, i.e. the one trained with PPO RL is simply an RNN that takes in the previous action as state + the internal state of the RNN, the only complication is that for each step they use a different softmax. For example in time steps 0, 1 the softmax is of size 25 then at 2, 3 it's 5 and finally at 4 it's 10. One can implement this using an RNN with multiple softmax outputting the correct softmax at each iteration by keeping an internal step and then reseting the step at the end of each episode. I was thinking of defining a custom model to do this, but before I take the hard path I wanted to ask. If I define the following 

```
agent = PPOAgent(
    states_spec=dict(type='float', shape=(1,)),
    actions_spec=environment.get_actions_spec(),
    network_spec=[
        dict(type='internal_lstm', size=100)
    ],
    batch_size=16,
    step_optimizer=dict(
        type='adam',
        learning_rate=1e-4
    )
)
```

where the actions are:

{'0': {'type': 'int', 'num_actions': 25}, '1': {'type': 'int', 'num_actions': 25}, '2': {'type': 'int', 'num_actions': 10}, '3': {'type': 'int', 'num_actions': 10}, '4': {'type': 'int', 'num_actions': 3}}

Is there a way to have the RNN output only at the correspoding softmax for each timestep and then do backprop properly? Please let me know. Thanks

**Comment by AlexKuhnle** [comment]
Assuming my understanding of the problem is correct (I would have to look at the paper to be sure), your approach sounds like a possible way of replicating their setup. Assuming that the actions 0 & 1 and 2 & 3 are supposed to be the same, semantically, you can reduce it to three actions. You could also introduce an additional no-op to each action and slightly punish not taking it when another action is required, but that's definitely not a great solution. Unfortunately it's not currently possible to represent this mechanism properly, and although we can put in on our to-do list, it won't happen in the next weeks. If you would want to look into integrating it yourself and contribute as a PR, we can probably give some guidance.

**Comment by AntreasAntoniou** [comment]
@AlexKuhnle Alright, can we have a chat about this somewhere online? I am very keen to get this done and yes I would be willing to implement features if they are needed, but if we can somehow avoid that even better. Where do we start?

**Comment by michaelschaarschmidt** [comment]
You can join our dev chat for discussion, filling in this form leads you there:
https://docs.google.com/forms/d/1_UD5Pb5LaPVUviD0pO0fFcEnx_vwenvuc00jmP2rRIc/edit

**Comment by trickmeyer** [comment]
@AntreasAntoniou @AlexKuhnle  any updates on this?  Can you link to the paper?  I'm currently trying to get this same thing (PPO w/ LSTM) going and am looking into how memory is sampled and if it's appropriate for PPO's optimizer.

**Comment by DaStoll** [comment]
Any udates on this? As I need this feature too I might be willing to help with/do a PR instead of writting my own RL implementation.

**Comment by AlexKuhnle** [comment]
Improved
